### ISSUE TITLE
OpenLinuxBoot meta-issue

### ISSUE THREAD

**Issue by mikebeaton** [acidanthera/bugtracker#1776]
[OpenLinuxBoot.efi](https://github.com/acidanthera/OpenCorePkg/pull/282) driver is now live: OC-native booting of Linux without chaining via GRUB.

The config.plist Drivers section layout has changed and must be updated (to allow driver arguments, see [sample files](https://github.com/acidanthera/OpenCorePkg/tree/master/Docs)).

Any queries or directly related issues may be discussed here.

The following work is still planned. The first ~should be done before 0.7.3 release~ is now fixed, the second (which is not proving needed in any distros seen so far, and is cosmetic not functional) will be done after 0.7.3 release:

TO FIX:

 - [x] Prevent over-long Linux boot entry names messing up layout of builtin menu (use ellipses when required)

TO ADD:

 - [x] Add support for `.contentDetails` and `.contentFlavour` files, to set entry base name and entry flavour, where auto-detection sets them wrong or finds no value. Currently auto-detection (incl. of `Flavour`, for known distros) is working well enough that this should not be required except in edge cases.

**Comment by vit9696** [comment]
Review DIFF: https://github.com/acidanthera/OpenCorePkg/compare/239c1a533d101c9e46c675a48585dbe548998582..aa1e19bebb20a5ad112fca31b111f8ad69cb8248

**Comment by vit9696** [comment]
https://github.com/acidanthera/OpenCorePkg/compare/239c1a533d101c9e46c675a48585dbe548998582..aa1e19bebb20a5ad112fca31b111f8ad69cb8248#diff-60f4b1e1b0708347f832a5adc65a2660a4f761d8ba3ffb95457523719dc85404R261
Extra spaces.

https://github.com/acidanthera/OpenCorePkg/compare/239c1a533d101c9e46c675a48585dbe548998582..aa1e19bebb20a5ad112fca31b111f8ad69cb8248#diff-fdea3738c880cc199119899414acd265da673bf245e57373b988f7c7dc510ce4R35
Use `#define GUID_STRING_LENGTH  36` from BaseLib.h

https://github.com/acidanthera/OpenCorePkg/compare/239c1a533d101c9e46c675a48585dbe548998582..aa1e19bebb20a5ad112fca31b111f8ad69cb8248#diff-b91b739de4873c914750ccace7d97a67b1fc09a82b08e8288389ad4b3e1d4896R37
Use do { } while (0)

https://github.com/acidanthera/OpenCorePkg/compare/239c1a533d101c9e46c675a48585dbe548998582..aa1e19bebb20a5ad112fca31b111f8ad69cb8248#diff-9b722fd38da2ff43af66ee2fe57135a111fdee1a3aadbcd06c8b3572fa151f63R1842
Boot argument parsing functions should go to a separate library (e.g. OcBootArgLib), along with the existing interfaces. The interfaces should probably be unified.

https://github.com/acidanthera/OpenCorePkg/compare/239c1a533d101c9e46c675a48585dbe548998582..aa1e19bebb20a5ad112fca31b111f8ad69cb8248#diff-6799435ff5ac1670c9068c2d4cd663645060c428d75b7a3af95fba4e71ad13bdR210
Should use CopyGuid here and below.

https://github.com/acidanthera/OpenCorePkg/pull/282/files/8cb3548f8e8b12ed85d28943f29e21d72621e35d#diff-fdea3738c880cc199119899414acd265da673bf245e57373b988f7c7dc510ce4
Might want to make OcXmlLib (are there any other?) use OcFlexArrayLib.

**Comment by cakehonolulu** [comment]
Hi!
I’ve been messing with OpenLinuxBoot for two days, checking the ins-n’-outs of it.

My major concern right now is:

The unability to use it on a SecureBoot-enabled environment.
Deleting the original signature from ext4_x64.efi using sbattach and resigning it with my own signature (Works for the rest of OpenCore); adding it to OC/Drivers with the proper entry in the config.plist yields all bootable entries unbootable (Black screen then back to OpenCanopy).

I’ve double-checked if disabling Secure Boot and using a bone-stock OpenCore Release (Non-debug) and OcBinaryData’s untouched ext4_x64.efi works and, yes, indeed does work.

What could the problem be? I’m willing to switch over Debug OpenCore if you need logs, I need Secure Boot enabled on my machine and unfortunately I can’t with OpenCore as it is (Because Grub chainloading doesn’t work anymore).

Thanks for your time!

**Comment by mikebeaton** [comment]
Chainloading grub does still work, nothing has been changed about that.

Yes, please do send a debug log with the signed ext4_x64.efi. Are you saying _all_ boot entries fail with this, or that all Linux boot entries fail. If I've understood correctly that it's just the latter, then your solution should be just to sign the individual Linux kernel images which you wish to boot `/boot/vmlinuz-{version}`. I believe that should just work, but equally you'll be the first person - to my knowledge - to test it.

**Comment by cakehonolulu** [comment]
> Chainloading grub does still work, nothing has been changed about that.
> 
> Yes, please do send a debug log with the signed ext4_x64.efi. Are you saying _all_ boot entries fail with this, or that all Linux boot entries fail. If I've understood correctly that it's just the latter, then your solution should be just to sign the individual Linux kernel images which you wish to boot `/boot/vmlinuz-{version}`. I believe that should just work, but equally you'll be the first person - to my knowledge - to test it.

Yes, all of them; which I find very strange.

As soon as I get home I’ll switch to OpenCore Debug and I’ll sign everything again.

The test I’ll do will consist in entering my Windows install, Ubuntu one and macOS one, maybe that’ll give some details!

**Comment by vit9696** [comment]
@cakehonolulu UEFI Secure Boot is still supported with Linux, you only need to add Linux certificates to your db variable. However, I additionally created https://github.com/acidanthera/bugtracker/issues/1796 for builtin UEFI Secure Boot support in the OpenCore.

**Comment by cakehonolulu** [comment]
> @cakehonolulu UEFI Secure Boot is still supported with Linux, you only need to add Linux certificates to your db variable. However, I additionally created #1796 for builtin UEFI Secure Boot support in the OpenCore.

Linux boots per-se on my regular Secure Boot config; no errors no nothing.

Last I checked Ubuntu chainloads GRUB using shimx64, which uses Microsoft CA keys, which I still mantain along my own ones (Wrapped both onto 1 so that I can maintain Windows too) so that shouldn’t be a problem I guess; but I can 100% be wrong so I’m open to corrections.

**Comment by cakehonolulu** [comment]
> Chainloading grub does still work, nothing has been changed about that.
> 
> Yes, please do send a debug log with the signed ext4_x64.efi. Are you saying _all_ boot entries fail with this, or that all Linux boot entries fail. If I've understood correctly that it's just the latter, then your solution should be just to sign the individual Linux kernel images which you wish to boot `/boot/vmlinuz-{version}`. I believe that should just work, but equally you'll be the first person - to my knowledge - to test it.

Just tested.

Signed all OC’s Debug components with my keys, Windows and macOS now work.

Ubuntu doesn’t work yet, complains about OCB: LoadImage failed - Security Violation.

My best guess is that something under Linux must be re-signed as per @vit9696 ’s recommendation. I’ll double check those signatures and come back with results.

**Comment by vit9696** [comment]
You are best to add Canonical keys to the db really. https://wiki.ubuntu.com/UEFI/SecureBoot/KeyManagement/ImageSigning. Their manual is really ugly as it does not show how to get their certificate, but here it is: https://git.launchpad.net/~ubuntu-core-dev/shim/+git/shim-signed/tree/CanonicalMasterCA.crt

**Comment by mikebeaton** [comment]
I've finally had a chance to set this up myself.

Of course (now that I have a chance to think about and explain it clearly!) the key issue is that OpenLinuxBoot never _boots_ anything - it just gives new boot entries to OpenCore, and the secure boot (and other) security on those entries is exactly the same as anything else OC boots; i.e., for a start, they won't ever boot if secure boot says they are invalid.

As to how to permit Linux kernels to be started directly (as they are with OpenLinuxBoot) in secure boot, I agree with Vit that the ideal is definitely to add the vendor-specific pub key to your secure boot `db` (image signing keys) variable - however, even with Vit's pointers above, I'm struggling to actually _find_ the vendor signing pub key for more than a very few distros (Ubuntu from Vit's link, Deepin from the shim-review issues, but so far not even a current - not too old and not too new for my recent install - key for Fedora, for example!).

My original suggestion of own-signing the kernel of course works, but (also of course) requires re-signing each new kernel after each update (so either booting via an old kernel, which OpenLinuxBoot will offer as an aux. entry, or booting with secure boot temporarily off).

If I can find more useful info about getting hold of valid, current distro kernel-signing pub keys, I'll link that here.

**Comment by mikebeaton** [comment]
Turns out I had a perfectly good Fedora cert, but `sbverify` refuses to verify it against Fedora kernels. Same tool which verifies other certs against other kernels in a seemingly reasonable way. But on installing the 'failed' cert in the SecureBoot `db` var, it allows Fedora to boot just fine. So unfortunately it seems `sbverify` tool is not completely reliable here - as if setting up secure boot wasn't fiddly enough already!

**Comment by mikebeaton** [comment]
Tool added to OC to extract vendor kernel signing cert from GRUB shim (e.g. shimx64.efi) file: https://github.com/acidanthera/OpenCorePkg/commit/e2819b2e3863b620bd7332c216852882f42956c9

This means there's no need to re-sign and there's a well defined way to get hold of the certificate required for the Secure Boot, for distros which ship with GRUB.

**Comment by eugene2878** [comment]
Hi. While .contentDetails and .contentFlavour features not implemented yet. How do you fix wrong icon choise by Openlinuxboot? I have few instances when Kubuntu distro shown as Ubuntu with Ubuntu icon and Kde neon using common Linux icon.

**Comment by mikebeaton** [comment]
Should be possible to change the icon with `.VolumeIcon.icns` file in root of Linux boot volume.

**Comment by mikebeaton** [comment]
> Hi. While .contentDetails and .contentFlavour features not implemented yet. How do you fix wrong icon choise by Openlinuxboot? I have few instances when Kubuntu distro shown as Ubuntu with Ubuntu icon and Kde neon using common Linux icon.

https://github.com/acidanthera/OpenCorePkg/commit/43137cc2775ad9a082e2b6b094c28eeba58fa1a0

**Comment by HJebbour** [comment]
I don't know if I should post my issue here, but when I boot from Ubuntu or Fedora from OC using OpenLinuxBoot my audio doesn't work (It shows Dummy Output and no Input for mic), When I boot from GRUB audio works properly both input and output, any particular reason why this issue happen only with audio when booting from OC?

**Comment by mikebeaton** [comment]
[Duplicate](https://github.com/acidanthera/OpenCorePkg/pull/282#issuecomment-1159503234) - already answered there.

**Comment by AlexFullmoon** [comment]
OpenLinuxBoot fails to process loader\entries\\\*.conf files on Fedora 41 due to Fedora adding some TuneD GRUB variables.

Specifically, in Fedora 40 conf files look like this:
```
title Fedora Linux (6.11.5-200.fc40.x86_64) 40 (Workstation Edition)
version 6.11.5-200.fc40.x86_64
linux /vmlinuz-6.11.5-200.fc40.x86_64
initrd /initramfs-6.11.5-200.fc40.x86_64.img
options root=UUID=f05a35ff-a897-4f55-ac04-5fcdf6db4dbf ro rootflags=subvol=root rhgb quiet
grub_users $grub_users
grub_arg --unrestricted
grub_class fedora
```
And in Fedora 41:
```
title Fedora Linux (6.11.5-300.fc41.x86_64) 41 (Workstation Edition)
version 6.11.5-300.fc41.x86_64
linux /vmlinuz-6.11.5-300.fc41.x86_64
initrd /initramfs-6.11.5-300.fc41.x86_64.img $tuned_initrd
options root=UUID=f05a35ff-a897-4f55-ac04-5fcdf6db4dbf ro rootflags=subvol=root rhgb quiet $tuned_params
grub_users $grub_users
grub_arg --unrestricted
grub_class fedora
```

Attached boot log, but here's the issue:
```
01:875 00:002 LNX: Reading 758c1891c87a4d4d947a61d4389e5ed3-6.11.5-300.fc41.x86_64.conf...
01:878 00:002 LNX: /initramfs-6.11.5-300.fc41.x86_64.img $tuned_initrd not found - Not Found
01:879 00:001 LNX: Error processing 758c1891c87a4d4d947a61d4389e5ed3-6.11.5-300.fc41.x86_64.conf - Not Found
```
[opencore-2024-11-01-070424.txt](https://github.com/user-attachments/files/17597136/opencore-2024-11-01-070424.txt)

This could be circumvented by removing those variables (as suggested [here](https://discussion.fedoraproject.org/t/kernel-install-92-tuned-install-is-conflicted-with-systemd-boot/134899)), but that would disable TuneD.

Not sure if I should post here or open separate issue.

**Comment by mikebeaton** [comment]
Thanks for the report. This should be a separate issue.
